language/stdlib/modules/association.move

```diff
@@ -109,6 +114,156 @@ module Association {
+        // TODO: Try to find a better approach to this that does not require excepting functions.
+        apply OnlyRootAddressHasRootPrivilege to *<T>, *
+            except initialize, root_address, has_privilege, addr_is_association,
+            assert_addr_is_association, assert_sender_is_association;
+    }
+    spec fun initialize {
+        // "initialize" establishes the invariant, so it's a special case.
+        // Before initialize, no addresses have a Root{} resource.
+        // Afterwards, only Root {} has the root resource.
+        requires all(addresses(), |a| !exists<Root>(a));
+        ensures root_invariant();
+    }
+
+    // This post-condition to assert_sender_is_root is a sanity check that
+    // the Root {} invariant really works. It needs the invariant
+    // OnlyRootAddressHasRootPrivilege, because it assert_sender_is_root does not
+    // directly check that the sender == root_address(). Instead, it aborts if
+    // sender has root privilege (Root{}), and only the root_address has Root{}
+    // TODO: There is a style question about whether this should just check for presence of
+    // a Root privilege. I guess it's moot so long as OnlyRootAddressHasRootPrivilege holds.
+    spec fun assert_sender_is_root {
+        ensures sender() == spec_root_address();
+    }
+
+    // ### authority to set is_certified flag
+
+    // SPEC: only grant_* functions can set the is_certified flag in
+    // PrivilegedCapability<Privilege>.
+    // Other functions can create and remove the Privilege object, so we have to condition
+    // everything on its existence.
+    spec schema OnlyGrantCanCertify<Privilege> {
+       ensures all(addresses(),
+                   |a| old(exists<PrivilegedCapability<Privilege>>(a)
+                           ==> !global<PrivilegedCapability<Privilege>>(a).is_certified)
+                        ==> (exists<PrivilegedCapability<Privilege>>(a)
+                             ==> !global<PrivilegedCapability<Privilege>>(a).is_certified));
+    }
+    spec module {
+        // By excepting only grant_*, we make sure only these two functions
+        // can change the is_certified from true to false.
+        // TODO: Try deleting the association version to see if prover catches it.
+        apply OnlyGrantCanCertify<Privilege> to *<Privilege>
+            except grant_privilege, grant_association_address;
+    }
+
+    // ### Functions can't make privileges disappear, except remove_privilege
+
+    // SPEC: Only remove_privilege can remove privileges
+    spec schema OnlyRemoveCanRemovePrivileges<Privilege> {
+         ensures any(addresses(), |a| (old(exists<PrivilegedCapability<Privilege>>(a))
+                                       && !exists<PrivilegedCapability<Privilege>>(a))
+                                       ==> sender() == spec_root_address());
+    }
+    spec module {
+        // FEATURE REQUEST: We need to be able to apply this to functions that don't have
```

Comment:
I think it would be clearer to instead allow quantification over types as part of the unbound quantification you already introduced for addresses, instead of letting the apply operator implicitly introduce one.

```
schema OnlyRemoveCanRemovePrivileges {
     invariant module all(types(), |t| ...t...)
}
```

That we need types as first class values is something Junkil already bumped on.

But this all depends on how well the unbound quantification actually works. From syntax viewpoint, once we have general unbounded quantification, we may also want to revisit the syntax.

Reply:
That seems like a good solution.